plugin/hooks.go

```diff
@@ -149,4 +154,41 @@ type Hooks interface {
 	// Note that this method will be called for files uploaded by plugins, including the plugin that uploaded the post.
 	// FileInfo.Size will be automatically set properly if you modify the file.
 	FileWillBeUploaded(c *Context, info *model.FileInfo, file io.Reader, output io.Writer) (*model.FileInfo, string)
+
+	// PushNotificationWillBeSent is invoked before the push notification has been sent to the push proxy server.
+	// You should return the notification that you want to be sent to the user (normally the same notification).
+	//
+	// You can reject the notification returning a nil notification.
+	//
+	// You can modify the notification, returning a changed version of the notification.
+	//
+	// Minimum server version: 5.10
+	PushNotificationWillBeSent(c *Context, notification *model.PushNotification) *model.PushNotification
+
+	// PushNotificationHasBeenSent is invoked after the push notification has been sent to the push proxy server.
+	//
+	// Minimum server version: 5.10
+	PushNotificationHasBeenSent(c *Context, notification *model.PushNotification)
+
+	// PushNotificationHasFailed is invoked after the push proxy server has replied with an error.
+	//
+	// Minimum server version: 5.10
+	PushNotificationHasFailed(c *Context, notification *model.PushNotification, err error)
+
+	// PushNotificationEnqueued is invoked just before the notification has been enqueued to be sent to the push proxy server.
+	//
+	// This is an internal API and it can change in the future.
+	//
+	// Minimum server version: 5.10
+	PushNotificationEnqueued(c *Context, notificationId, notificationType, userId, channelId, postId string)
+
+	// PushNotificationAck is invoked when the server receive the confirmation
+	// of the reception of the push notification from the mobile device.
+	//
+	// You can return data that will be return to the mobile device as response to the ACK.
```

Comment:
Could you give an example on what kind of data could be returned?

Is this really something plugin authors can modify, as the client would need to know what do with the data?

Reply:
I going to change it to a notification, because is the unique data type that makes sense. If the client receive a notification from here, update the current notification in the app (or at least, the interanal representation, to for example, open the correct channel and post if you have removed it from the original notification). This is because this way you can make a plugin which remove every single bit of internal information from the notification (so this way is not passing through google/apple servers, and only when it arrives the other end, you can get the information from the server using this ACK endpoint).